contributors/design-proposals/predicates-ordering.md

```diff
@@ -0,0 +1,90 @@
+
+
+
+## Static ordering
+
+This ordering will be the default ordering. If a policy config is provided with a subset of predicates, only those predicates will be invoked using the static ordering. 
+
+
+
+
+|Position                  | Predicate                        | comments (notes, justifications...)              |
+ ----------------- | ---------------------------- | ------------------
+| 1 | `CheckNodeConditionPredicate`  | we really don’t wanna check predicates against unschedulable nodes. |
+| 2           | `PodFitsHost`            | we check the pod.spec.nodeName. |
+| 3           | `PodFitsHostPorts` | we check ports asked on the spec. |
+| 4 | `PodMatchNodeSelector`            | check node label after narrowing search. |
+| 5           | `PodFitsResources `            | this one comes here since it’s not restrictive enough as we do not try to match values but ranges. |
+| 6           | `NoDiskConflict` | Following the resource predicate, we check disk |
+| 7 | `PodToleratesNodeTaints `            | check toleration here, as node might have toleration |
+| 8          | `PodToleratesNodeNoExecuteTaints`            | check toleration here, as node might have toleration |
+| 9           | `CheckNodeLabelPresence ` | labels are easy to check, so this one goes before |
+| 10 | `checkServiceAffinity `            | - |
+| 11           | `MaxPDVolumeCountPredicate `            | - |
+| 12           | `VolumeNodePredicate ` | - |
+| 13 | `VolumeZonePredicate `            | - |
+| 14           | `CheckNodeMemoryPressurePredicate`            | doesn’t happen often |
+| 15           | `CheckNodeDiskPressurePredicate` | doesn’t happen often |
+| 16 | `InterPodAffinityMatches`            | Most expensive predicate to compute |
+
+
+## End-user ordering
+
+using scheduling policy file, the cluster admin can override the default static ordering. This gives administrator the maximum flexibility regarding scheduler behaviour and enables scheduler to adapt to cluster usage. as an example the following is scheduler policy file using an end-user ordering:
+
+``` json
+{
+"kind" : "Policy",
+"apiVersion" : "v1",
+"predicates" : [
+	{"name" : "PodFitsHostPorts", "order": 2},
+	{"name" : "PodFitsResources", "order": 3},
+	{"name" : "NoDiskConflict", "order": 5},
+	{"name" : "PodToleratesNodeTaints", "order": 4},
+	{"name" : "MatchNodeSelector", "order": 6},
+	{"name" : "PodFitsHost", "order": 1}
+	],
+"priorities" : [
+	{"name" : "LeastRequestedPriority", "weight" : 1},
+	{"name" : "BalancedResourceAllocation", "weight" : 1},
+	{"name" : "ServiceSpreadingPriority", "weight" : 1},
+	{"name" : "EqualPriority", "weight" : 1}
+	],
+"hardPodAffinitySymmetricWeight" : 10
+}
+```
+
+
+## Timeline
+
+both types of ordering will be delivered as an alpha version, as for the timeline:
+
+* static ordering: available in 1.9
```

Comment:
Static ordering is fine to do straight away, as it doesn't change how users use clusters, but I think we probably should discuss putting dynamic ordering initially as beta. I'm not sure if it's worth it - we need to discuss it with SIG PM (whatever they handle is).

Reply:
@gmarek - I thought that we'd wait for metrics from the community to start working on the dynamic ordering ?

Reply:
I think it is a good idea to add static ordering first and wait for feedback from users before adding dynamic ordering. If users asked for dynamic ordering, we will add it.

Reply:
@gmarek @bsalamat - I'm a bit confused here, dynamic ordering isn't part of this design proposal, since we need data on predicates restrictiveness.

I'm fine with just having for static ordering and the end-user one for the moment.

Reply:
I was referring to configurable ordering by end-users as "dynamic ordering". Do we already know customers who need the configurable ordering?

Reply:
I still would like to hear who are the customers who needed it and had to write their own custom scheduler only to reorder predicates and why a smarter ordering (that we are proposing in static ordering of this proposal) is not enough for them.

Reply:
If you are using very specific set of labels, executing checkNodeLabelPresence first is better. 

@gmarek - I'll try to join the next SIG PM meeting to discuss this.

Reply:
Oh, sorry, if the 'static' means 'at startup' and dynamic means 'changes according to the cluster state' then you're right, I understood it differently. It's probably not worth joining the meeting, as they most likely have full agenda, email should work better.

@apsinha @timothysc - do you know if we should use normal alpha-beta-GA (most likely skipping alpha) lifecycle for things like this (allowing user to change the order of predicate computation in scheduler - setting good one improves scheduler's performance, setting terrible one pretty much kills it). It's small enough change from the user perspective, that maybe we can skip this?